core/block/editor/chatobject/chatobject.go

```diff
@@ -110,11 +126,188 @@ func (s *storeObject) Init(ctx *smartblock.InitContext) error {
+		return
+	}
+	txn, err := coll.WriteTx(s.componentCtx)
+	if err != nil {
+		log.With(zap.Error(err)).Error("markReadMessages: start write tx")
+		return
+	}
+	ctx := txn.Context()
+	idsModified := make([]string, 0, len(ids))
+	for _, id := range ids {
+		if id == s.Id() {
+			// skip tree root
+			continue
+		}
+		res, err := coll.UpdateId(ctx, id, query.MustParseModifier(`{"$set":{"`+readKey+`":true}}`))
+		if err != nil {
+			log.With(zap.Error(err)).With(zap.String("id", id)).With(zap.String("chatId", s.Id())).Error("markReadMessages: update message")
+			continue
+		}
+		if res.Modified > 0 {
+			idsModified = append(idsModified, id)
+		}
+	}
+	err = txn.Commit()
+	if err != nil {
+		log.With(zap.Error(err)).Error("markReadMessages: commit")
+		return
+	}
+	log.Debug(fmt.Sprintf("markReadMessages: %d/%d messages marked as read", len(idsModified), len(ids)))
+	if len(idsModified) > 0 {
+		// it doesn't work within the same transaction
+		// query the new oldest unread message's orderId
+		iter, err := coll.Find(
+			query.Key{Path: []string{readKey}, Filter: query.NewComp(query.CompOpEq, false)},
+		).Sort(ascOrder).
+			Limit(1).
+			Iter(s.componentCtx)
+		if err != nil {
+			log.With(zap.Error(err)).Error("markReadMessages: failed to find oldest unread message")
+		}
+		defer iter.Close()
+		var newOldestOrderId string
+		if iter.Next() {
+			val, err := iter.Doc()
+			if err != nil {
+				log.With(zap.Error(err)).Error("markReadMessages: failed to get oldest unread message")
+			}
+			if val != nil {
+				newOldestOrderId = val.Value().GetObject(orderKey).Get("id").GetString()
+			}
+		}
+		log.Debug(fmt.Sprintf("markReadMessages: new oldest unread message: %s", newOldestOrderId))
+		s.subscription.chatState.Messages.OldestOrderId = newOldestOrderId
+		s.subscription.updateReadStatus(idsModified, true)
+		s.onUpdate()
+	}
+}
+
+func (s *storeObject) MarkReadMessages(ctx context.Context, afterOrderId, beforeOrderId string, lastAddedMessageTimestamp int64) error {
+	// 1. select all messages with orderId < beforeOrderId and addedTime < lastDbState
+	// 2. use the last(by orderId) message id as lastHead
+	// 3. update the MarkSeenHeads
+	// 2. mark messages as read in the DB
+
+	msg, err := s.GetLastAddedMessageInOrderRange(ctx, afterOrderId, beforeOrderId, lastAddedMessageTimestamp)
+	if err != nil {
+		return fmt.Errorf("get message: %w", err)
+	}
+
+	// mark the whole tree as seen from the current message
+	s.storeSource.MarkSeenHeads([]string{msg.Id})
+	return nil
+}
+
+func (s *storeObject) GetLastAddedMessageInOrderRange(ctx context.Context, afterOrderId, beforeOrderId string, lastAddedMessageTimestamp int64) (*model.ChatMessage, error) {
+	coll, err := s.store.Collection(ctx, collectionName)
+	if err != nil {
+		return nil, fmt.Errorf("get collection: %w", err)
+	}
+
+	if lastAddedMessageTimestamp < 0 {
+		// todo: remove this
+		// for testing purposes
+		lastAddedMessageTimestamp = math.MaxInt64
+	}
+	iter, err := coll.Find(
```

Comment:
Should we also add filter for unread messages?

Reply:
fixed